### PR TITLE
changed the K visitor to use KORE.

### PR DESCRIPTION
**Pull request by grosu**
The previous visitor was too complex to translate automatically into KORE, so I manually changed it into one using directly KORE.  The new visitor can still be improved, but it did not lose anything from the previous one and it can now be easily translated/handled.

### CODE REVIEW

**Comment by grosu** [comment]
Traian, can you please review it?

BTW, Dwight, I cannot find in the Wiki page explanations on how to contact a reviewer for your code in a way that is recorded.  Should we do that outside github, or via a comment like I am just doing it here?

File: tutorial/README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # K Tutorial
```
Comment (traiansf) [inline tutorial/README.md:1]: This file probably should not have been part of this patch.

File: include/modules/k-visitor.k [hunk 2]
```diff
@@ -1,56 +1,44 @@
-require "builtins/k-equal.k"
```
Comment (traiansf) [inline include/modules/k-visitor.k:1]: Consider adding JavaDoc to explain at least the exported functionality.
At least turn some comments into javadoc comments.

**Comment by dwightguth** [comment]
Grigore: the standard way to request that someone reviews their code is to tag them in the description or a comment of the review: for example, "@traiansf please review."

File: include/modules/k-visitor.k [hunk 3]
```diff
@@ -1,56 +1,44 @@
...
+  syntax KLabel ::= "#visitor"
+  syntax KLabel ::= "#visited"
+
+// #visitor takes an action and a predicate, and applies the action on each
+// visited KItem that satisfies the predicate.  The action is typically a
+// curried funtion, and the predicate is just a function (encoded as a label item).
+  rule #visit(L:KLabel(Kl:KList),, #visitor(A:KLabel(Kla:KList),, #label(P)))
+    => #if P(L(Kl)) ==K true
+       #then A(Kla,,L(Kl))
+       #else #visiting.kapp(#label(L),, #visit(Kl,, #visitor(A(Kla),, #label(P))))
```
Comment (traiansf) [inline include/modules/k-visitor.k]: For uniformity, P should be allowed to take arguments, too.

Comment (traiansf) [inline include/modules/k-visitor.k]: I maintain my opinion that A(Kla) should be written as `#klabel(A), #kList(Kla)`, but I would accept it as A(Kla), too.

**Comment by grosu** [comment]
@traiansf Please review.

I have added all the changes that you suggested, except for removing the trivial change on the tutorial/README.md file from the patch, because that is not such a big deal anyway.  The likelihood to mess things up while reverting and jumping from branch to branch are higher in my case than just letting that line be part of this push request.

BTW, I added LOTS of comments to quote/unquote, which reminded me that you actually added no comments to Agent at all.  When you get a chance, please do so.  Use the quote/unquote as an example.

File: include/modules/k-visitor.k [hunk 4]
```diff
@@ -1,56 +1,107 @@
+// Copyright @ 2012-2014.  K Team.  All Rights Reserved.
+
 require "builtins/k-equal.k"
 
+/*@ \subsection{Generic Visitors}
+
+\K allows to define generic visitors, that is, visitors that take any K term
+and transform it into according to given parameters.
```
Comment (traiansf) [inline include/modules/k-visitor.k]: perhaps remove "into"

File: include/modules/k-visitor.k [hunk 5]
```diff
@@ -1,56 +1,107 @@
...
+The particular user-defined syntax of the target language is irrelevant for
+the visitor, because that is all eventually transformed into K terms, and the
+visitors work with the latter.
+This is in sharp constrast with conventional visitor-like semantic approaches,
+such as those envountered in conventional semantics of
```
Comment (traiansf) [inline include/modules/k-visitor.k]: encountered.  Actually, please spell check.  There are a couple of misspelled words in the following.

File: include/modules/k-visitor.k [hunk 6]
```diff
@@ -1,56 +1,107 @@
...
+\begin{itemize}
+\item \texttt{\#visit}, which is expected to take two arguments, the K term
+to visit and the actual visitor.
+\item \texttt{\#visitor}, which is a \texttt{KItem} stating what to do during
+the visit.  Currently, \texttt{\#visitor} expects four arguments:
+the first two describe the action to take, and the later two describe the
+condition under which to take the action.
+We need two arguments for each because both the action and the condition
+consist of a \texttt{KLabel} and a partial list of arguments for it.
+The complete list of argument is obtained by appending the actual node being
```
Comment (traiansf) [inline include/modules/k-visitor.k]: arguments

File: include/modules/k-visitor.k [hunk 7]
```diff
@@ -1,56 +1,107 @@
...
-       #else visiting Kl:KList applying A:KLabel if Pred:KLabel 
+  syntax KItem  ::= #label(KLabel)       // Should be eventually builtin
+  syntax KLabel ::= "#klist"             // Should be predefined, with many others
+
+//@ This is the API of the generic visitor:
+  syntax KLabel ::= "#visit"
+  syntax KLabel ::= "#visitor"
+  syntax KLabel ::= "#visited"
+
+/*@ \texttt{\#visitor} takes an action and a predicate, and \texttt{\3visit}
```
Comment (traiansf) [inline include/modules/k-visitor.k]: #visit

**Comment by grosu** [comment]
@traiansf Thanks, Traian.  I fixed the typos, so I think we are done with this unless you have more comments.  Are you OK with how the visitors are now?  That is, the 4 arguments?  I know you probably don't like the fact that I have #visited(Kl:Klist) instead of #visited(#klist(Kl)), but this is more of a conceptual issue than technical, and we can change it later if we really want to.

**Comment by dwightguth** [comment]
retest this please

**Comment by dwightguth** [comment]
retest this please

**Comment by dwightguth** [comment]
@grosu: Your tests are now failing because of the change I committed which requires all modified files in a commit to be updated with copyright headers. Hooray!

**Comment by grosu** [comment]
Please let me know what I can do to help, but you know my knowledge about git :), so do not have high expectations from me.

**Comment by dwightguth** [comment]
All you have to do is download the resulting error log from Jenkins and fix all the errors. You can also test it yourself in your working copy by running:

git diff master --name-only > .checkstyle-files
ant checkstyle-copyright

I will put this on the Wiki.

**Comment by grosu** [comment]
Which wiki is it?  We have lots of them there, we should probably do some cleanup.  I had copyright notes in my two files, namely

// Copyright @ 2012-2014.  K Team.  All Rights Reserved. 

Should I just replace @ with (C)?

Grigore

**Comment by dwightguth** [comment]
Well it's a bit of a finicky regular expression right now. We can talk
about loosening it maybe, but what it accepts right now would be:

// Copyright (C) 2012-2014 K Team. All Rights Reserved.

Note also the lack of period after the years. It's whitespace- and
case-sensitive and everything. But I think being strict is okay here
because all the copyright notices I've seen e.g. at Google were the same
way. We can discuss tuning the whitespace also, but I like it the way it is
because it has punctuation roughly the same way most copyright declarations
in the front of books do. I also chose (C) because @ isn't really the
copyright symbol.

I agree we should clean up the wikis some, but I added this information to
the Testing page because it governs tests passing/failing, which seemed
logical to me.

**Comment by grosu** [comment]
Done.  I prefer it strict, too.

File: samples/agent/quote-unquote.k [hunk 8]
```diff
@@ -1,35 +1,110 @@
...
+
+/*@ \subsubsection{Syntax}
+
+\texttt{lift} and \texttt{eval} are strict, where the former takes the
+resulting value and lifts it into a code value, and the later expects its
+argument to evaluate to a code value and turns it into its coresponding
+code, which is consequently evaluated in the current context.
+\texttt{quote} and \texttt{unquote} are not strict.
+The former freezes its argument code into a code value, without evaluating
+it.  Exception makes subparts of the code which appear as arguments of
```
Comment (traiansf) [inline samples/agent/quote-unquote.k]: I don't think I understand this sentence.

File: samples/agent/quote-unquote.k [hunk 9]
```diff
@@ -1,35 +1,110 @@
...
+/*@ Finally, we define the auxilliary predicate testing if a code fragment is a
+\texttt{quote} or \texttt{unquote}:  */
+  syntax Bool ::= isQuoted(Exp)     [function]
+  rule isQuoted(quote E) => true
+  rule isQuoted(unquote K) => true
 
-  rule isVal(visitedK(_)) => true
+/*@  Conceptually, the above is the conventional definition of
+\texttt{quote}/\texttt{unquote}.
+However but the definitions that we encountered so far were all language
```
Comment (traiansf) [inline samples/agent/quote-unquote.k]: please revise "However but"

**Comment by traiansf** [comment]
Ok, I'm fine with it. Other improvements can be done later (e.g., making the labels actually strict once we fix #312)
Pleas feel free to merge it once Jenkins passes all tests, unless you want somebody more English proficient than me to look over the comments.

**Comment by grosu** [comment]
Fixed.  Yes, it wouldn't hurt if Dwight takes a look, too, and fixes the obvious English mistakes.

**Comment by dwightguth** [comment]
Grigore: There is a bug in my Jenkins code that I am tracking down right now which is causing these tests to fail even after you fixed the copyright. I will probably fix this yet today, and then take a look at the rest of it tomorrow morning.

**Comment by dwightguth** [comment]
Okay, I think I fixed the issue with Jenkins not diffing against the right set of files. I will have to monitor the next couple pull requests people try to create to be sure, but when I play around with the commit tree it behaves like I expect it to, so I think I probably fixed it.

As I said, I will get to the rest of this review tomorrow morning first thing.

File: include/modules/k-visitor.k [hunk 10]
```diff
@@ -1,56 +1,107 @@
+// Copyright (C) 2012-2014 K Team. All Rights Reserved.
+
 require "builtins/k-equal.k"
 
+/*@ \subsection{Generic Visitors}
+
+\K allows to define generic visitors, that is, visitors that take any K term
```
Comment (dwightguth) [inline include/modules/k-visitor.k]: "allows users to" is a little more ordinary English

File: include/modules/k-visitor.k [hunk 11]
```diff
@@ -1,56 +1,107 @@
...
+and transform it according to given parameters.
+The particular user-defined syntax of the target language is irrelevant for
+the visitor, because that is all eventually transformed into K terms, and the
+visitors work with the latter.
+This is in sharp constrast with conventional visitor-like semantic approaches,
+such as those encountered in conventional semantics of
+\texttt{quote}/\texttt{unquote} conmstructs in languages with support for code
+generation, which are language-specific, that is, which have a case (smeantic rule)
```
Comment (dwightguth) [inline include/modules/k-visitor.k]: s/smeantic/semantic/

File: samples/agent/quote-unquote.k [hunk 12]
```diff
@@ -1,35 +1,109 @@
...
+inside the argument expression of a \texttt{quote}).
+Since the result of applying a visitor to a K term is the visited K term wrapped
+with label \texttt{\#visited}, the semantics of the constructs \texttt{lift}
+and \texttt{eval} are defined using the \texttt{\#visited} wrapper: */
+
+  syntax KItem ::= quoteit(Int,K)
+  rule quote E => mQuote(E, 0)
+  rule quoteit(N, quote E)   => #visiting.kapp(#label('quote_),,mQuote(E, N +Int 1))
+  rule quoteit(0, unquote E) => E
+  rule quoteit(N, unquote E) => #visiting.kapp(#label('unquote_),,mQuote(E, N +Int 1))
```
Comment (dwightguth) [inline samples/agent/quote-unquote.k]: should this be -Int 1? If so, this code doesn't seem to work right now. Why isn't it failing some kind of test?

File: samples/agent/quote-unquote.k [hunk 13]
```diff
@@ -1,35 +1,109 @@
...
+The \texttt{unquote} construct is expected to produce a code value, otherwise
+a runtime error will occur if executed (because \texttt{unquote} can only occur
+inside the argument expression of a \texttt{quote}).
+Since the result of applying a visitor to a K term is the visited K term wrapped
+with label \texttt{\#visited}, the semantics of the constructs \texttt{lift}
+and \texttt{eval} are defined using the \texttt{\#visited} wrapper: */
+
+  syntax KItem ::= quoteit(Int,K)
+  rule quote E => mQuote(E, 0)
+  rule quoteit(N, quote E)   => #visiting.kapp(#label('quote_),,mQuote(E, N +Int 1))
```
Comment (dwightguth) [inline samples/agent/quote-unquote.k:82]: If we intend to expose #visiting.kapp to the user, I think we need to document somewhere in k-visitor.k what this use case is.

**Comment by dwightguth** [comment]
Looks really good! I'd have to play around with it more to know if it would meet my visitor needs in Python, but it seems a lot cleaner and less confusing now. I had some small comments that I wanted to address before this is merged though... see above.

**Comment by grosu** [comment]
@traiansf @dwightguth  please review and merge

Took care of all your comments now.

**Comment by traiansf** [comment]
Updates to k-visitor and quote/unquote look good to me.

**Comment by grosu** [comment]
Who is rv-jenkins?  Dwight?

Grigore

**Comment by dwightguth** [comment]
Grigore: I didn't want our rvmonitor code to be tested by a github user named kframework-jenkins so I created a new bot user for the Jenkins machine in the company office. Unfortunately it got a little overzealous and tried to build the kframework builds also. I think I should have taken care of this for now, but we need to talk at the k meeting today about which box to use to test the k framework.

**Comment by dwightguth** [comment]
I still want to see some kind of test case for quote-unquote that fails if the semantics has the bug that I found. Otherwise I think we probably aren't tested enough.

**Comment by grosu** [comment]
Hm, I thought I took care of it: It is Traian's job to do that, because it is his language, Agent.  I've already done a lot more than I should, by commenting his code :)

Grigore

**Comment by traiansf** [comment]
Apparently there are no tests run for AGENT, as the test job for it looks
for programs with extension .agent, while the AGENT programs traditionally
did not have any extension (as the could be run by partial instances of the
modular AGENT language).

I can probably bring them back into the circuit, but I don't think I can do
it in this branch.  So, how should we proceed?

2014-04-08 21:33 GMT+03:00 Grigore Rosu notifications@github.com:

**Comment by dwightguth** [comment]
Okay, I'm going to merge this then. Please make fixing the tests for AGENT a high priority todo for you then.

**Comment by grosu** [comment]
Can you please put an issue about it, so we keep track of it?

Grigore

**Pull request merged**